rfc/store-deliver.md

```diff
@@ -0,0 +1,310 @@
+# extend store/* protocol with store/deliver 
+
+## Authors
+
+- [Vasco Santos], [Protocol Labs]
+
+## Abstract
+
+`store/*` protocol is missing verifiability for content actually uploaded to the provided presigned URL. This factor together with intention to create a proof of delivery makes critical the introduction of a new capability in the `store/*` namespace. In this RFC, we propose `store/deliver` and `store/confirm` capabilities.
+
+## Introduction
+
+When an agent wants to store some content, agent invokes `store/add` to declare this intention. The `store/add` invocation MUST include the CID of the content to be stored. An invocation example:
+
+```json
+{
+  "can": "store/add",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "size": 1234
+  }
+}
+```
+
+If the service is receptive to store this data, a presigned URL is returned in the receipt of `store/add` invocation, so that the client can store this content. The presigned URL includes a hash derived from the content CID to guarantee the written bytes match that CID. A receipt example:
+
+```json
+{
+  "ran": "bafy...storeAdd",
+  "out": {
+    "ok": {
+      "status": "upload",
+      "url": "https://...",
+      "link": "bag...",
+      "allocated": 1234
+    }
+  },
+  "fx": {
+    "join": null,
+    "fork": []
+  },
+}
+```
+
+Note that content requested to store, may already be stored. If that is the case, the service may issue receipt with `"status": "done"`.
+
+On their schedule, the agent can post the bytes using the presigned URL. However, both the agent and the service never sign a receipt stating that this content is now stored. This is where `store/deliver` and `store/confirm` can help. In short, we can see `store/deliver` as a way for the agent to sign that they did the work, and `store/confirm` for the service to sign that the bytes were received, stored and can now be retrieved.
+
+While not relevant for verifiability, it is also important pointing out that the service MAY need to run some data computation, issue billing events or write indexes for retrievability asynchronously. Today, web3.storage operated store service relies on "S3 Bucket events" to achieve this, but it should not be assumed that all stores web3.storage will rely on will provide such features.
+
+## New capabilities
+
+### `store/deliver` capability
+
+After the agent handled the asynchronous task of posting the bytes, the agent can invoke `store/deliver` to notify the service that the bytes were delivered. This invocation in the future MAY include a challenge for the service based on the posted data.
+
+```json
+{
+  "can": "store/deliver",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "url": "https://..."
+  }
+}
+```
+
+### `store/confirm` capability
```

Comment:
We really should have this 

1. In a different name space
2. Make resource service DID not user did.

Because it is service that can invoke this not the user and user has ownership of their space key not us.

Reply:
yeah, I think resource here should probably be a did of the write target?